### PR TITLE
[fix](recycler) Fix CountdownEvent error and hang

### PR DESCRIPTION
**Pull request by Yukang-Lian**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

Fix CountdownEvent error "Invoking add_count() after wait() was invoked"

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [x] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [x] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [x] No.
    - [ ] Yes. 

- Does this need documentation?
    - [x] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by Thearas** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

**Comment by Yukang-Lian** [comment]
run buildall

File: cloud/src/recycler/sync_executor.h [hunk 1]
```diff
@@ -70,6 +70,7 @@ class SyncExecutor {
         return res;
     }
     void reset() {
```
Comment (gavinchou) [inline cloud/src/recycler/sync_executor.h:72]: also fix where it is called or consider call it in `when_all()`

**Comment by Yukang-Lian** [comment]
run buildall

File: cloud/src/recycler/sync_executor.h [hunk 2]
```diff
@@ -65,11 +66,21 @@ class SyncExecutor {
                 *finished = false;
                 return res;
             }
-            res.emplace_back((*task).get());
+            auto status = task->wait_for(10);
```
Comment (gavinchou) [inline cloud/src/recycler/sync_executor.h:69]: comment, what is the magic

Comment (gavinchou) [reply cloud/src/recycler/sync_executor.h:69]: syncpoint change to wait timeout and test

**Comment by Yukang-Lian** [comment]
run buildall

File: cloud/src/recycler/sync_executor.h [hunk 3]
```diff
@@ -65,11 +69,27 @@ class SyncExecutor {
...
                 return res;
             }
-            res.emplace_back((*task).get());
+            // 10s
+            size_t max_wait_time = 10000;
+            TEST_SYNC_POINT_CALLBACK("SyncExecutor::when_all.set_wait_time", &max_wait_time);
+            // _count.timed_wait has already ensured that all tasks are completed.
+            // The 10 seconds here is just waiting for the task results to be returned,
+            // so 10 seconds is more than enough.
+            auto status = task->wait_for(max_wait_time);
```
Comment (gavinchou) [inline cloud/src/recycler/sync_executor.h:78]:
```suggestion
            size_t max_wait_ms = 10000;
            TEST_SYNC_POINT_CALLBACK("SyncExecutor::when_all.set_wait_time", &max_wait_ms);
            // _count.timed_wait has already ensured that all tasks are completed.
            // The 10 seconds here is just waiting for the task results to be returned,
            // so 10 seconds is more than enough.
            auto status = task->wait_for(max_wait_ms);
```

**Review by gavinchou — APPROVED** [review]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by bobhan1 — APPROVED** [review]

**Pull request merged**